private/bufpkg/bufcheck/buflint/internal/buflintvalidate/cel.go

```diff
@@ -0,0 +1,311 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package buflintvalidate
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/bufbuild/buf/private/gen/proto/go/buf/validate"
+	"github.com/bufbuild/buf/private/pkg/protosource"
+	"github.com/bufbuild/protovalidate-go/celext"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+const (
+	// https://buf.build/bufbuild/protovalidate/docs/main:buf.validate#buf.validate.FieldConstraints
+	celFieldTagInFieldConstraints = 23
+	// https://buf.build/bufbuild/protovalidate/docs/main:buf.validate#buf.validate.MessageConstraints
+	celFieldTagInMessageConstraints = 3
+)
+
+var (
+	messageConstraintsExtensionType = dynamicpb.NewExtensionType(validate.E_Message.TypeDescriptor())
+	fieldConstraintsExtensionType   = dynamicpb.NewExtensionType(validate.E_Field.TypeDescriptor())
```

Comment:
Definitely worth comments as to what these are here for, since `validate.E_Message` and `validate.E_Field` already implement `protoreflect.ExtensionType`.

My understanding (from other conversations) is that they need to be _dynamic_ extensions to avoid the type-assertion-panic that might otherwise happen when querying their value from the results of compilation.

You have `getReflectMessageDescriptor` in here, so you already have a way to get the `protoreflect.MessageDescriptor` that corresponds to a particular `protosource.Message`. Why not instead use that instead _with_ the protovalidate runtime to avoid having to repeat so much of what's in protovalidate here?

Reply:
As another benefit of using `protovalidate-go`, these two variables are no longer needed. Resolved in [59e3caf](https://github.com/bufbuild/buf/pull/2473/commits/59e3caf2b1e1b4f024c52a17ec8372ce6e6b051e)